rfcs/0007-handling-corruption.md

```diff
@@ -0,0 +1,533 @@
+# Handling Data Corruption
+
+Status: Draft
+
+Authors:
+
+* [Derrick J Wippler](https://github.com/thrawn01)
+
+References:
+
+- https://www.sqlite.org/atomiccommit.html
+- https://avi.im/blag/2024/sqlite-bit-flip/
+- https://github.com/danthegoodman1/BreakingSQLite
+- https://www.sqlite.org/recovery.html
+- https://www.postgresql.org/docs/9.1/app-pgresetxlog.html
+- https://www.postgresql.org/docs/current/app-initdb.html#APP-INITDB-DATA-CHECKSUMS
+- https://www.postgresql.org/docs/current/runtime-config-developer.html#GUC-IGNORE-CHECKSUM-FAILURE
+- https://www.postgresql.org/docs/current/runtime-config-developer.html#GUC-ZERO-DAMAGED-PAGES
+- https://docs.tigerbeetle.com/about/safety/
+- https://github.com/thrawn01/db-corruption
+
+## Motivation
+The durability of object storage is a significant factor driving the adoption
+of disaggregated storage. However, we should not assume that data corruption
+will never occur. Several factors can lead to corruption, including:
+
+- Programming errors in SlateDB
+- Malicious intent to corrupt data
+- Variations in the logical durability of different object store implementations
+- Errors in firmware and drivers
+
+How SlateDB operates in the face of corruption is crucial for its adoption and
+for maintaining operator confidence.
+
+Although SlateDB utilizes object stores that typically provide some level of
+Logical Availability and Durability, SlateDB should be viewed as a
+manifestation of a physical system. This is because SlateDB cannot make
+assumptions about the Logical Availability or Durability of the object store it
+uses to store data. The object store provided to SlateDB at startup could
+simply be a local disk masquerading as a highly available logical store.
+
+### Prioritizing Durability
+Prioritizing durability means that in the event of data corruption, the
+database will immediately become unavailable until the corruption is repaired
+through some form of out-of-band intervention. This approach ensures that the
+database cannot lose data without the operator's knowledge.
+
```

Comment:
I think there's a partial availability option here as well. In partial availability, the DB fails all operations that attempt to read corrupt data, but operations that don't read corrupt data continue to function normally. For example, if the SST with `k1` in it is corrupt, and a subsequent (more recently written) SST has `k2`, if I were to read `k2`, I would get back a its value. But if I were to read `k1`, I would get an error. Essentially, rather than fully disabling the DB, only fail on operations where corrupt data is seen.

Reply:
I assumed that If a newer version of `k1` exists in a corrupt SST, but an older version exists in an uncorrupted SST, the database will break it's consistency guarantees. This might be a faulty assumption, but I wanted to error on the side of caution. If you choose availability, you are also accepting the possibility of inconsistent behavior (implied by lost data) in order to remain available.

I'll update the RFC to make "lost data implies inconsistent behavior" clear.

Reply:
I think I'm saying something slightly different. In my example `k1` and `k2` are different keys. So the DB could return the latest value of `k2` since its SST is more recent than `k1`'s. This would allow partial availability without any consistency issues on corrupt keys (they'd still return a corruption error).

Reply:
I follow what you are saying now.

I'm changing the wording to indicate that corruption does not result in database closure, only that the current operation will return an error. The operation IE: `get` can be preformed on another key, if that key lookup doesn't encounter corruption it will succeed without error. This allows for "partial availability" even when prioritizing durability over availability.